kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java

```diff
@@ -71,7 +73,19 @@ public CompletionStage<RequestFilterResult> onProduceRequest(short apiVersion,
                                                                  ProduceRequestData request,
                                                                  FilterContext context) {
         return maybeEncodeProduce(request, context)
-                .thenCompose(yy -> context.forwardRequest(header, request));
+                .thenCompose(yy -> context.forwardRequest(header, request))
+                .exceptionallyCompose(throwable -> {
+                    if (isEncryptionException(throwable) || isEncryptionException(throwable.getCause())) {
```

Comment:
I guess this is making the assumption that all encryption exceptions are transient. I.e. that letting the client retry is always the correct strategy. 

If we think that's right (and will always be right) then we should document that fact in the Javadoc of `EncryptionException`. 

Otherwise we could model it more explicitly somehow. For example we could use a boolean flag parameter in the `EncryptionException` constructor, forcing ourselves to always choose at the point we throw an `EncryptionException`. Or have EE return an `ApiException` cause, to decide what the error response code should be. Or use a exception type hierarchy....

Probably the Javadoc is enough for now.

Reply:
>I guess this is making the assumption that all encryption exceptions are transient. I.e. that letting the client retry is always the correct strategy.

>If we think that's right (and will always be right) then we should document that fact in the Javadoc of EncryptionException.

That's a good point, yes it does assume/force all encryption exceptions to be treated as transient. I think that's inherent in being a transparent proxy. 

> Otherwise we could model it more explicitly somehow. For example we could use a boolean flag parameter in the EncryptionException constructor, forcing ourselves to always choose at the point we throw an EncryptionException. Or have EE return an ApiException cause, to decide what the error response code should be. Or use a exception type hierarchy....

Out of those options I'd probably chose 
>forcing ourselves to always choose at the point we throw an EncryptionException
And making at a constructor argument along side message and cause.

I'll add some javadoc with a call to action to talk to us if that assumption is problematic.

Reply:
Well crap https://github.com/kroxylicious/kroxylicious/blob/574f67316e69831f259df689f12865d922ab2ffb/kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/crypto/WrapperV1.java#L60 thats a non transient usage. 

I'll add the ApiException as property then.